### PR TITLE
Synchronize ZeRO-3 rollout early stopping

### LINKED ISSUE #8321: HybridEngineRollout: synchronized early stopping across ZeRO-3 data-parallel ranks
## Description

Follow-up to #8264 (tracked there in the review discussion, agreed to be handled separately).

To keep ZeRO-3 data-parallel ranks in lockstep during rollout, `HybridEngineRollout.generate` currently passes `eos_token_id=None` to the underlying generate call and restores EOS/padding semantics after the fact via `_pad_after_eos`. This fixes the deadlock, but as a consequence **every rank always decodes the full `max_new_tokens`**, even when every sequence in the global batch has already produced its EOS. The wasted compute is significant when typical responses are much shorter than the max response length.

## Proposal

All ranks should collectively detect the early-stop condition — every sequence on every rank's batch has produced EOS — and stop the decode loop together, while preserving the ZeRO-3 invariant that all ranks execute the same number of parameter-gather collectives.

Concretely, the stop decision must be *global*, not local:

- Periodically (every decode step, or every K steps) all-reduce a local "unfinished sequences remain" flag across data-parallel ranks.
- All ranks stop only when the aggregated flag says no rank has unfinished sequences. Any rank exiting the loop based on local knowledge alone would reintroduce the deadlock #8264 fixed.

This can be implemented either as a custom decode loop with an explicit sync point, or as a `StoppingCriteria` that aggregates the decision globally before returning `True` on all ranks.

## Workload context (from #8264)

- Main 8-GPU OPSD run: `micro_batch_size_per_gpu=1`, `gradient_accumulation_steps=1`, `n_samples_per_prompt=1` → 8 prompts / 8 rollouts per training iteration across 8 DP ranks.
- Verification runs use max response length 64; full reproduction uses 1024. Responses are typically far shorter than the max, so decoding to `max_new_tokens` on every rank wastes a large fraction of rollout compute.

## Design consideration

With `micro_batch_size_per_gpu=1` there is exactly one sequence per GPU, so per-slot work reclamation (continuous batching) has nothing to reclaim locally — synchronized early stopping *is* the whole win for this configuration. Continuous batching becomes interesting for larger per-rank batches.

Please implement the stop-check as a reusable synchronized control-flow primitive (e.g. a periodic "global work remaining" rendezvous) rather than a one-off stop hack: continuous batching generalizes the same rendezvous from a "stop/continue" boolean into a "rebuild the batch composition" decision, so an early-stopping fix designed with that shape becomes a stepping stone instead of rework.

@LiRunGuo — you kindly offered to own this follow-up in #8264. GitHub won't let me assign you until you've commented on this issue (non-collaborator limitation), so please drop a comment here and I'll assign it to you.

### PR DESCRIPTION
**Pull request by LiRunGuo**
## Summary

- synchronize EOS stopping decisions across the ZeRO-3 parameter-gather process group
- stop at the latest rank-local EOS instead of always decoding to `max_new_tokens`
- retain each sequence's first EOS and pad later response positions
- keep non-ZeRO-3 generation on the native EOS path
- disable the `synced_gpus` default that Transformers derives from the default world size on the ZeRO-3 path

The rendezvous uses `optimizer.dp_process_group`, which is the group used by ZeRO-3 parameter gathers, rather than assuming the default world group. CUDA/HIP graph-capture rollout remains outside this change because it bypasses Hugging Face `generate()`.

## Why `synced_gpus` must be disabled

When Transformers can observe ZeRO-3 it defaults `synced_gpus` to True from the default world size, and `_has_unfinished_sequences()` then issues an **ungrouped world all-reduce**. A data-parallel subgroup that enters rollout while other WORLD ranks stay out would block on that collective; concurrently active subgroups would also be coupled unnecessarily. The group-aware criterion already gives every rank in the parameter-gather group the same stop decision, so the ZeRO-3 path passes `synced_gpus=False` and the non-ZeRO-3/FSDP paths keep the Transformers default.

Closes #8321.

## Validation

- DeepSpeed pre-commit checks: passed on the changed files
- 4-rank CPU/Gloo two-subgroup test: passed
- 4-rank CPU/Gloo subgroup generation regression (HF-visible ZeRO-3 via `HfDeepSpeedConfig`, ranks 2-3 never enter rollout): passes with the fix, does not complete without it
- 2-rank AMD MI300X scripted-EOS ZeRO-3 test: ranks reaching EOS at decode steps 2 and 5 both executed 5 forwards (`[5, 5]`), while preserving local response lengths
- 2-GPU Qwen2.5-0.5B/7B ZeRO-3 OPSD smoke test: completed two steps with finite losses (`0.6784`, `0.1554`)

Exact validation head: `1a0e951b9e3e09b83bfe10c4de7e5358495f6664`.

### CODE REVIEW

File: deepspeed/runtime/rollout/hybrid_engine_rollout.py [hunk 1]
```diff
@@ -69,6 +70,36 @@ def _elapsed_time(self, start, end):
...
+        self.eos_token_ids = torch.as_tensor(eos_token_id).flatten()
+        self.work_remaining = work_remaining
+
+    def __call__(self, input_ids, scores, **kwargs):  # noqa: ARG002
+        response_ids = input_ids[:, self.response_start:]
+        eos_token_ids = self.eos_token_ids.to(device=input_ids.device, dtype=input_ids.dtype)
+        sequence_has_eos = (response_ids.unsqueeze(-1) == eos_token_ids).any(dim=-1).any(dim=-1)
+        local_work_remaining = (~sequence_has_eos).any().to(dtype=torch.int32)
+        global_work_remaining = self.work_remaining(local_work_remaining)
+        return torch.full((input_ids.shape[0], ), not global_work_remaining, dtype=torch.bool, device=input_ids.device)
```
[the remark this answers is not in the document]
Comment (LiRunGuo) [reply deepspeed/runtime/rollout/hybrid_engine_rollout.py]: Fixed in d8a5b1779. The stopping criterion now returns a scalar Python bool because the decision is already rank-global. This preserves the intended all-local-sequences behavior, is compatible with Transformers 4.32 (which applies Python any() to criterion results), and also broadcasts correctly in newer StoppingCriteriaList implementations. I added a two-sequence regression assertion covering the 4.32-style any() path.

File: tests/unit/runtime/rollout/test_hybrid_engine_rollout.py [hunk 2]
```diff
@@ -9,18 +9,25 @@
 """
 
 from types import SimpleNamespace
+from types import MethodType
 from unittest.mock import MagicMock, patch
 
 import pytest
 import torch
+from torch import distributed as torch_dist
```
[the remark this answers is not in the document]
Comment (LiRunGuo) [reply tests/unit/runtime/rollout/test_hybrid_engine_rollout.py]: Fixed in d8a5b1779. I removed the direct torch.distributed import and initialize/query process groups only through deepspeed.comm. Since this is specifically a CPU/Gloo subgroup topology test, the test temporarily mocks DeepSpeed's accelerator as CPU during deepspeed.init_distributed(); this avoids binding Gloo to a visible GPU on accelerator hosts while retaining the DeepSpeed communication abstraction. The 4-rank subgroup test and the 2-GPU scripted-EOS ZeRO-3 test both pass in Slurm job 400417.

**Comment by FU-max-boop** [comment]
@LiRunGuo I found one remaining subgroup liveness issue at exact head `d8a5b1779eafcbf15ab1a14e04aa83bf04f75c3e`, so I am holding the requested A100 run for now.

`HybridEngineRollout.generate()` installs the new `optimizer.dp_process_group`-aware stopping criterion but leaves `synced_gpus` unset. When the Transformers `HfDeepSpeedConfig` integration can see ZeRO-3, Transformers auto-enables `synced_gpus` from the default world size in both [4.32.1](https://github.com/huggingface/transformers/blob/v4.32.1/src/transformers/generation/utils.py#L1395-L1399) and [4.51.3](https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/generation/utils.py#L2229-L2230). Its decode loop then performs an ungrouped/default-world all-reduce in [4.32.1](https://github.com/huggingface/transformers/blob/v4.32.1/src/transformers/generation/utils.py#L2430-L2437) and [4.51.3](https://github.com/huggingface/transformers/blob/v4.51.3/src/transformers/generation/utils.py#L2607-L2615). If only one DP subgroup enters rollout, that first world collective can wait forever for non-participating ranks; concurrently active subgroups are also unnecessarily coupled through WORLD.

The current 4-rank Gloo test calls `GlobalWorkRemaining` directly, so it does not enter Transformers `generate()`. The 2-rank CUDA test uses the full WORLD and initializes with `deepspeed.initialize()` directly, which ordinarily does not establish the Transformers ZeRO-3 weakref, so it likely does not trigger the implicit `synced_gpus=True` path either.

The narrow fix is to pass `synced_gpus=False if zero3_sync is not None else None`, preserving the Transformers default for non-ZeRO-3/FSDP paths. The custom criterion already gives every rank in the parameter-gather group the same stop decision. I would also add a kwargs assertion plus a generate-level regression with HF-visible ZeRO-3 and a proper subgroup/non-participating WORLD ranks. This is source-level and CPU/Gloo-testable; once a fixed head lands, NVIDIA validation can target that exact commit.

**Comment by LiRunGuo** [comment]
@FU-max-boop thanks for the precise analysis — you are right that leaving `synced_gpus` unset would make this worse, not better, once Transformers can see ZeRO-3.

I rebased onto current `master` (the branch was conflicting after the continuous-batching work) and addressed the liveness issue at head `1a0e951b9`:

- `generate()` now passes `synced_gpus=False if zero3_sync is not None else None`, so the ZeRO-3 path opts out of the world-size-derived default while non-ZeRO-3/FSDP paths keep the Transformers behavior.
- Added a kwargs assertion (`test_zero3_generation_disables_native_synced_gpus`) that pins `synced_gpus is False` together with `eos_token_id is None` on the ZeRO-3 path, and asserts the non-ZeRO-3 path still leaves `synced_gpus` unset.
- Added `TestZero3SubgroupGeneration` (world_size=4, Gloo): ZeRO-3 is made visible to Transformers via `HfDeepSpeedConfig`, ranks 0-1 generate inside one `dp_process_group` subgroup while ranks 2-3 never enter rollout. Without the change the test does not complete (world all-reduce waits on the non-participating ranks); with the change it passes.

Local validation on this head (CPU/Gloo, torch 2.9.1+rocm6.4, transformers 5.15.0):

- `pytest tests/unit/runtime/rollout/test_hybrid_engine_rollout.py -k TestZero3SubgroupGeneration` → `1 passed` (~61s)
- same test with `synced_gpus` reverted to the Transformers default → hangs, never completes
- `pytest ... -k "not Subgroup and not Zero3SynchronizedEarlyStop ..."` → `32 passed`
- `pre-commit run --files deepspeed/runtime/rollout/hybrid_engine_rollout.py tests/unit/runtime/rollout/test_hybrid_engine_rollout.py` → all hooks pass

The PR is mergeable again. This head is the one to target for the A100 run whenever you have a slot. Happy to add anything else you want covered before that.

**Comment by LiRunGuo** [comment]
@FU-max-boop friendly follow-up: the `synced_gpus` liveness fix you asked for is in `1a0e951b9` (details in my comment above), including a 4-rank subgroup regression test that hangs without the change. Could you re-check it, and run the A100 validation on this head when you have a slot?

@tohtana could you take a look too when you have time? The related #8392 (ZeRO-3 QKV layout persistence, +58/−2) is also still waiting for a first review.
